src/evaluateXPath.ts

```diff
@@ -136,6 +149,101 @@ const evaluateXPath = <TNode extends Node, TReturnType extends keyof IReturnType
 		throw new TypeError("Failed to execute 'evaluateXPath': xpathExpression must be a string.");
 	}
 
+	const backend = options && options.backend ? options.backend : 'expression';
+
+	if (backend === 'js-codegen' || backend === 'auto') {
+		const compiledJavaScriptResult = compileWithJsCodegenBackend(selector, returnType, options);
+
+		if (compiledJavaScriptResult.isAstAccepted) {
+			const wrappedDomFacade: DomFacade = new DomFacade(
+				domFacade === null ? new ExternalDomFacade() : domFacade
+			);
+			const contextArray = adaptJavaScriptValueToArrayOfXPathValues(
+				wrappedDomFacade,
+				contextItem,
+				'item()?'
+			);
+			return compiledJavaScriptResult.result.evaluate(contextArray[0], wrappedDomFacade);
+		} else if (backend !== 'auto') {
+			// When only using the js-codegen backend, we should not proceed if
+			// it fails.
+			throw new Error(
+				`Failed compiling the given XPath with the js-codegen backend. ${compiledJavaScriptResult.reason}`
+			);
+		}
+	}
+
+	try {
+		return evaluateWithExpressionBackend(
+			selector,
+			contextItem,
+			domFacade,
+			variables,
+			returnType,
+			options
+		);
+	} catch (error) {
+		printAndRethrowError(selector, error);
+	}
+};
+
+const compileWithJsCodegenBackend = <
+	TNode extends Node,
+	TReturnType extends keyof IReturnTypes<TNode>
+>(
+	selector: string,
+	returnType?: TReturnType,
+	options?: Options | null
+): CompiledJavaScriptResult => {
+	const expressionString = normalizeEndOfLines(selector);
+
+	const compilationOptions = {
+		allowXQuery:
+			options['language'] === Language.XQUERY_3_1_LANGUAGE ||
+			options['language'] === Language.XQUERY_UPDATE_3_1_LANGUAGE,
+		debug: !!options['debug'],
+		disableCache: !!options['disableCache'],
+	};
+
+	let compiledJavaScriptResult = compilationOptions.disableCache
+		? null
+		: getCompiledJavaScriptFromCache(expressionString, returnType);
+
+	if (!compiledJavaScriptResult) {
+		const ast = parseExpression(expressionString, compilationOptions);
+		const mainModule = astHelper.getFirstChild(ast, 'mainModule');
+
+		if (!mainModule) {
+			// This must be a library module
+			throw new Error('Can not execute a library module.');
```

Comment:
Also here, don't you wanna reject instead of throwing? Now you're likely missing the special XPath error code.